### PR TITLE
workflows/eval: add markdown of added, removed and changed

### PR DESCRIPTION
**Pull request by itepastra**
Makes the github action for eval output the added, removed and changed paths in the summary.

The paths were already available in json format in `changed-paths.json`.
This PR adds a human-readable representation to the summary of the format (not in a code block)
```
## Added


## Removed

graalvmCEPackages.graalvm-ce-musl
graalvmCEPackages.stdenv

## Changed

babashka
babashka-unwrapped
bbin
certificate-ripper
clj-kondo
cljfmt
clojure-lsp
cq
dapl-native
dbqn-native
graalvm-ce
graalvmCEPackages.graaljs
graalvmCEPackages.graalnodejs
graalvmCEPackages.graalpy
graalvmCEPackages.graalvm-ce
graalvmCEPackages.truffleruby
jet
neil
obb
scala-update
tests.writers.bin.babashka
tests.writers.simple.babashka
tests.writers.wrapping.babashka
tests.writers.wrapping.babashka-bin
vscode-extensions.betterthantomorrow.calva
yamlscript
zprint
```

## Things done

created https://github.com/itepastra/nixpkgs/pull/3 to see what the result looks like

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Mic92** [comment]
Do you have a test PR, what this looks like? Just make a pull request against your `ci-eval-markdown` branch.

**Comment by itepastra** [comment]
https://github.com/itepastra/nixpkgs/actions/runs/12095524145/attempts/1#summary-33728887151 <- this is what it currently looks like

File: ci/eval/default.nix [hunk 1]
```diff
@@ -261,6 +261,12 @@ let
           --slurpfile after ${afterResultDir}/outpaths.json \
           > $out/changed-paths.json
 
+        echo -e "## Added\n" > $out/step-summary.md
+        jq -r '.attrdiff.added[]' $out/changed-paths.json >> $out/step-summary.md
+        echo -e "\n## Removed\n" >> $out/step-summary.md
+        jq -r '.attrdiff.removed[]' $out/changed-paths.json >> $out/step-summary.md
+        echo -e "\n## Changed\n" >> $out/step-summary.md
+        jq -r '.attrdiff.changed[]' $out/changed-paths.json >> $out/step-summary.md
```
[the remark this answers is not in the document]
Comment (Mic92) [reply ci/eval/default.nix:269]: This makes it an enumeration.

[the remark this answers is not in the document]
Comment (itepastra) [reply ci/eval/default.nix]: I didn't know this was a thing jq could do, I just thought "why not try the simplest thing I can make first". I've updated the commit

**Review by infinisil — APPROVED** [review]
Thanks, LGTM!

File: ci/eval/default.nix [hunk 2]
```diff
@@ -261,6 +261,11 @@ let
           --slurpfile after ${afterResultDir}/outpaths.json \
           > $out/changed-paths.json
 
+        jq -r '
+        "## Added\n" + (.attrdiff.added | map("- \(. )") | join("\n")) + "\n" +
+        "## Removed\n" + (.attrdiff.removed | map("- \(. )") | join("\n")) + "\n" +
+        "## Changed\n" + (.attrdiff.changed | map("- \(. )") | join("\n")) + "\n"
```
Comment (azuwis) [inline ci/eval/default.nix:267]:
```suggestion
        "## Added " + (.attrdiff.added | length | tostring) + "\n" + (.attrdiff.added | map("- \(. )") | join("\n")) + "\n" +
        "## Removed " + (.attrdiff.removed | length | tostring) + "\n" + (.attrdiff.removed | map("- \(. )") | join("\n")) + "\n" +
        "## Changed " + (.attrdiff.changed | length | tostring) + "\n" + (.attrdiff.changed | map("- \(. )") | join("\n")) + "\n"
```

How about this, will produce something like:

---

## Added 0

## Removed 0

## Changed 25
- hello
- tests.checkpointBuildTools
- tests.config.allowPkgsInPermittedInsecurePackages
- tests.devShellTools.nixos
- tests.devShellTools.unstructuredDerivationInputEnv
- tests.devShellTools.valueToString
- tests.hardeningFlags.stackClashProtectionExplicitDisabled
- tests.hardeningFlags.stackClashProtectionExplicitEnabled
- tests.hardeningFlags.stackClashProtectionStdenvUnsupp
- tests.hardeningFlags-clang.stackClashProtectionExplicitDisabled
- tests.hardeningFlags-clang.stackClashProtectionExplicitEnabled
- tests.hardeningFlags-clang.stackClashProtectionStdenvUnsupp
- tests.hardeningFlags-gcc.stackClashProtectionExplicitDisabled
- tests.hardeningFlags-gcc.stackClashProtectionExplicitEnabled
- tests.hardeningFlags-gcc.stackClashProtectionStdenvUnsupp
- tests.pkgs-lib
- tests.testers.nixosTest-example
- tests.testers.runNixOSTest-example
- tests.testers.testBuildFailure.helloDoesNotFail
- tests.trivial-builders.concat
- tests.trivial-builders.linkFarm
- tests.trivial-builders.references
- tests.trivial-builders.writeClosure-union
- tests.trivial-builders.writeShellApplication
- tests.trivial-builders.writeStringReferencesToFile

**Comment by azuwis** [comment]
Some experiments:

```
$ jq -r '.rebuildsByKernel | to_entries[] | "* \(.key) \(.value | length | tostring)", "  * \(.value[])"' < changed-paths.json
* darwin 4
  * hello
  * tests.checkpointBuildTools
  * tests.config.allowPkgsInPermittedInsecurePackages
  * tests.devShellTools.nixos
* linux 6
  * hello
  * tests.checkpointBuildTools
  * tests.config.allowPkgsInPermittedInsecurePackages
  * tests.devShellTools.nixos
  * tests.devShellTools.unstructuredDerivationInputEnv
  * tests.devShellTools.valueToString
```

* darwin 4
  * hello
  * tests.checkpointBuildTools
  * tests.config.allowPkgsInPermittedInsecurePackages
  * tests.devShellTools.nixos
* linux 6
  * hello
  * tests.checkpointBuildTools
  * tests.config.allowPkgsInPermittedInsecurePackages
  * tests.devShellTools.nixos
  * tests.devShellTools.unstructuredDerivationInputEnv
  * tests.devShellTools.valueToString

**Comment by azuwis** [comment]
```
$ jq -r '.attrdiff | to_entries[] | "## \(.key | ascii_upcase) \(.value | length | tostring)", "- \(.value[])"'
## ADDED 0
## REMOVED 0
## CHANGED 5
- hello
- tests.checkpointBuildTools
- tests.config.allowPkgsInPermittedInsecurePackages
- tests.devShellTools.nixos
- tests.devShellTools.unstructuredDerivationInputEnv
```

Can produce very similar result to https://github.com/NixOS/nixpkgs/pull/360339#discussion_r1864254042

---

## ADDED 0
## REMOVED 0
## CHANGED 25
- hello
- tests.checkpointBuildTools
- tests.config.allowPkgsInPermittedInsecurePackages
- tests.devShellTools.nixos
- tests.devShellTools.unstructuredDerivationInputEnv

**Comment by azuwis** [comment]
Update https://github.com/NixOS/nixpkgs/pull/360339#discussion_r1864254042:

- Use `to_entries`, less repetition
- Include both `attrdiff` and `rebuildsByKernel`
- Use nested list
 
See if you like it or not :D

File: .github/workflows/eval.yml [hunk 3]
```diff
@@ -210,7 +210,7 @@ jobs:
             --arg beforeResultDir ./baseResult \
             --arg afterResultDir ./prResult \
             -o comparison
-
+          cat comparison/step-summary.md >> "$GITHUB_STEP_SUMMARY"
```
Comment (itepastra) [inline .github/workflows/eval.yml:213]: Oh, I lost this line somewhere? Thanks for fixing it

**Comment by Mic92** [comment]
@azuwis I took the idea of adding numbers to each section, but also added a truncate function as step summaries have a 1MB size limit. Because of the size limitation, I would for now not include other section for now and just stick to `added/removed/changed` for now.

**Comment by Mic92** [comment]
I now also added links to the github search: https://gist.github.com/Mic92/b3f1200a4b92535c4dc3ecde9bab6019

**Comment by infinisil** [comment]
Manual backport: https://github.com/NixOS/nixpkgs/pull/361168

**Comment by infinisil** [comment]
Whoops we have [a problem](https://github.com/NixOS/nixpkgs/actions/runs/12287784044/job/34290600043#step:9:18):

> Error: $GITHUB_STEP_SUMMARY upload aborted, supports content up to a size of 1024k, got 1636k. For more information see: https://docs.github.com/actions/using-workflows/workflow-commands-for-github-actions#adding-a-markdown-summary

**Comment by Mic92** [comment]
To be honest, do we want those markdown files anyway for mass-rebuilds?
Maybe we can just skip them. Because there is not much point in listing all packages of nixpkgs.

**Comment by Mic92** [comment]
Maybe we just truncate the document with `sed`?

**Comment by Mic92** [comment]
Actually I already did a limit function there. So maybe we just need to adjust the limit to a safe value?

**Comment by infinisil** [comment]
@Mic92 Oh yeah that sounds good

**Comment by infinisil** [comment]
Ideally it also shows how to get the full list

**Pull request merged**